consensus/ledger_validator/ledger_validator.c

```diff
@@ -85,44 +88,58 @@ static retcode_t build_snapshot(ledger_validator_t *const lv,
-    //         if
-    //         (Snapshot.isConsistent(milestoneTracker.latestSnapshot.patchedDiff(
-    //                 stateDiffViewModel.getDiff()))) {
-    //           milestoneTracker.latestSnapshot.apply(stateDiffViewModel.getDiff(),
-    //                                                 candidateMilestone.index());
-    *consistent_index = milestone.index;
-    memcpy(consistent_hash, milestone.hash, FLEX_TRIT_SIZE_243);
-    //         } else {
-    //           break;
-    //         }
-    //       }
+    if ((ret = iota_tangle_state_delta_load(lv->tangle, milestone.index,
+                                            &delta)) != RC_OK) {
+      goto done;
+    }
+    if (delta != NULL && !state_delta_empty(delta)) {
+      if ((ret = iota_snapshot_create_patch(
+               lv->milestone_tracker->latest_snapshot, &delta, &patch)) !=
+          RC_OK) {
+        goto done;
+      }
+      if (state_delta_is_consistent(&patch)) {
+        if ((ret = iota_snapshot_apply_patch(
+                 lv->milestone_tracker->latest_snapshot, &delta,
+                 milestone.index)) != RC_OK) {
+          goto done;
+        }
+        *consistent_index = milestone.index;
+        memcpy(consistent_hash, milestone.hash, FLEX_TRIT_SIZE_243);
+      } else {
+        break;
+      }
+    }
     hash_pack_reset(&pack);
     if ((ret = iota_tangle_milestone_load_next(lv->tangle, milestone.index,
                                                &pack)) != RC_OK) {
       goto done;
     }
+    state_delta_destroy(&delta);
+    state_delta_destroy(&patch);
   }
 
 done:
+  rw_lock_handle_unlock(&lv->milestone_tracker->latest_snapshot->rw_lock);
+  state_delta_destroy(&delta);
+  state_delta_destroy(&patch);
   return ret;
 }
 
-static retcode_t get_latest_delta(ledger_validator_t *const lv,
+static retcode_t get_latest_delta(ledger_validator_t const *const lv,
                                   hash_set_t *const analyzed_hashes,
-                                  state_delta_t *state, flex_trit_t *const tip,
-                                  uint64_t latest_snapshot_index,
-                                  bool is_milestone, bool *valid_delta) {
+                                  state_delta_t *const state,
+                                  flex_trit_t const *const tip,
+                                  uint64_t const latest_snapshot_index,
+                                  bool const is_milestone,
+                                  bool *const valid_delta) {
   retcode_t ret = RC_OK;
-  int number_of_analyzed_transactions = 0;
   hash_stack_t non_analyzed_hashes = NULL;
+  hash_set_t counted_hashes = NULL;
   iota_transaction_t tx_bundle = NULL;
   DECLARE_PACK_SINGLE_TX(tx, tx_ptr, pack);
-  struct _trit_array tx_hash = {NULL, NUM_TRITS_HASH, FLEX_TRIT_SIZE_243, 0};
+  struct _trit_array curr_hash = {NULL, NUM_TRITS_HASH, FLEX_TRIT_SIZE_243, 0};
+  flex_trit_t null_hash[FLEX_TRIT_SIZE_243];
```

Comment:
this should be defined as static and initialized outside the context of this function,
we have several places already where we have that (`genesis_hash`)
maybe add a struct with definitions that will be initialized via `consensus_init`?

Reply:
We have `common/defs.h`, I can add `common/defs.c` as well for variables like that.